### ISSUE TITLE
Custom rules (`user_rules`) not run when also specifying a `rules` allow list with the `Linter` class

### ISSUE THREAD

**Issue by Samyak2** [sqlfluff/sqlfluff#3974]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I'm trying out sqlfluff to use as a base to implement my custom linting rules. So I implemented a class inheriting from `BaseRule` and tried to use it with the `Linter` class and the `user_rules` parameter.

Then I found a built-in rule that I also want to enable. When I did so using the `rules` parameter of the `Linter` class, the custom rule I had specified in `user_rules` did not run,

### Expected Behaviour

Both the rules I enabled and the custom rules are run.

### Observed Behaviour

Only the enabled built-in rules are run. The custom rule is not run.

### How to reproduce

Here's an example:
```python
from typing import List

from sqlfluff.core import Linter
from sqlfluff.core.rules import BaseRule, LintResult, RuleContext
from sqlfluff.core.rules.crawlers import RootOnlyCrawler


class Rule_Something_L001(BaseRule):
    """Does nothing."""

    groups = ("all",)

    crawl_behaviour = RootOnlyCrawler()

    def _eval(self, context: RuleContext) -> List[LintResult]:
        print("hello world im here")
        return []


linter = Linter(user_rules=[Rule_Something_L001], dialect="ansi")  # relevant line
result = linter.lint_string("SELECT 1")
```

This outputs `hello world im here` as expected.

But when I change the linter initialization to:
```python
linter = Linter(user_rules=[Rule_Something_L001], dialect="ansi", rules=["L001"])
```

I get no output, showing that my custom rule was not run.

---

I could not figure out the root cause from a cursory exploration of the `Linter` and `FluffConfig` classes.

### Dialect

ANSI

### Version

1.3.2

### Configuration

None

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by Samyak2** [comment]
Update: this is not the right way to add a custom rule, I need to make a plugin instead as answered here: https://github.com/sqlfluff/sqlfluff/issues/3973#issuecomment-1285354935
